platform/kube/inject/initializer.go

```diff
@@ -0,0 +1,456 @@
+			i.clientset.ExtensionsV1beta1().RESTClient(),
+			&v1beta1.DaemonSet{},
+			i.initializeDaemonSet,
+		},
+		{
+			"replicasets",
+			i.clientset.ExtensionsV1beta1().RESTClient(),
+			&v1beta1.ReplicaSet{},
+			i.initializeReplicaSet,
+		},
+		{
+			"replicationcontrollers",
+			i.clientset.CoreV1().RESTClient(),
+			&v1.ReplicationController{},
+			i.initializeReplicationController,
+		},
+		{
+			"pods",
+			i.clientset.CoreV1().RESTClient(),
+			&v1.Pod{},
+			i.initializePod,
+		},
+	}
+
+	for _, kind := range kinds {
+		kind := kind // capture the current value of `kind`
+
+		watchList := cache.NewListWatchFromClient(kind.getter, kind.resource,
+			i.options.Namespace, fields.Everything())
+
+		// Wrap the returned watchlist to workaround the inability to include
+		// the `IncludeUninitialized` list option when setting up watch clients.
+		includeUninitializedWatchList := &cache.ListWatch{
+			ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+				options.IncludeUninitialized = true
+				return watchList.List(options)
+			},
+			WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+				options.IncludeUninitialized = true
+				return watchList.Watch(options)
+			},
+		}
+
+		_, controller := cache.NewInformer(includeUninitializedWatchList, kind.objType, i.options.ResyncPeriod,
+			cache.ResourceEventHandlerFuncs{
+				AddFunc: func(obj interface{}) {
+					if err := kind.initialize(obj); err != nil {
+						glog.Errorf("Could not initialize %s: %v", kind.resource, err)
+					}
+				},
+			},
+		)
+		i.controllers = append(i.controllers, controller)
+	}
+	return i
+}
+
+func (i *Initializer) initializerPresent(objectMeta *metav1.ObjectMeta) bool {
+	glog.V(2).Infof("ObjectMeta initializer info %v/%v policy:%q status:%q %v",
+		objectMeta.Namespace, objectMeta.Name,
+		objectMeta.Annotations[istioSidecarAnnotationPolicyKey],
+		objectMeta.Annotations[istioSidecarAnnotationStatusKey],
+		objectMeta.Initializers)
+
+	if objectMeta.GetInitializers() == nil {
+		return false
+	}
+	pendingInitializers := objectMeta.GetInitializers().Pending
+	if len(pendingInitializers) == 0 {
+		return false
+	}
+	if initializerName != pendingInitializers[0].Name {
+		return false
+	}
+
+	return true
+}
+
+func (i *Initializer) modifyResource(objectMeta *metav1.ObjectMeta, templateObjectMeta *metav1.ObjectMeta, spec *v1.PodSpec) error { // nolint: lll
+	switch i.options.Namespace {
+	case v1.NamespaceAll:
+		// skip kubernetes system namespace
+		if objectMeta.Namespace == "kube-system" {
+			return nil
+		}
+	case objectMeta.Namespace:
+	default:
+		// Skip namespace(s) that we're not responsible for
+		// initializing.
+		return nil
+	}
+
+	// Remove self from the list of pending Initializers while
+	// preserving ordering.
+	if pending := objectMeta.GetInitializers().Pending; len(pending) == 1 {
+		objectMeta.Initializers = nil
+	} else {
+		objectMeta.Initializers.Pending = append(pending[:0], pending[1:]...)
```

Comment:
This was carried over from keley's initializer tutorial. It's a type'd `nil` which avoids reallocating the underlying array (see https://github.com/golang/go/wiki/SliceTricks#filtering-without-allocating)